src/track/taglib/trackmetadata_file.cpp

```diff
@@ -0,0 +1,188 @@
+
+namespace mixxx {
+
+namespace {
+
+Logger kLogger("TagLib");
+
+} // anonymous namespace
+
+namespace taglib {
+
+// Deduce the TagLib file type from the file name
+FileType getFileTypeFromFileName(
+        const QString& fileName) {
+    DEBUG_ASSERT(!fileName.isEmpty());
+    const QString fileExt(fileName.section(QChar('.'), -1).toLower().trimmed());
+    if (QStringLiteral("mp3") == fileExt) {
+        return FileType::MP3;
+    }
+    if (QStringLiteral("m4a") == fileExt) {
+        return FileType::MP4;
+    }
+    if (QStringLiteral("flac") == fileExt) {
+        return FileType::FLAC;
+    }
+    if (QStringLiteral("ogg") == fileExt) {
+        return FileType::OGG;
+    }
+    if (QStringLiteral("opus") == fileExt) {
+        return FileType::OPUS;
+    }
+    if (QStringLiteral("wav") == fileExt) {
+        return FileType::WAV;
+    }
+    if (QStringLiteral("wv") == fileExt) {
+        return FileType::WV;
+    }
+    if (fileExt.startsWith(QStringLiteral("aif"))) {
+        return FileType::AIFF;
+    }
+    return FileType::Unknown;
+}
+
+QDebug operator<<(QDebug debug, FileType fileType) {
+    return debug << static_cast<std::underlying_type<FileType>::type>(fileType);
+}
+
+bool hasID3v1Tag(TagLib::MPEG::File& file) {
+#if (TAGLIB_HAS_TAG_CHECK)
+    return file.hasID3v1Tag();
+#else
+    return file.ID3v1Tag() != nullptr;
+#endif
+}
+
+bool hasID3v2Tag(TagLib::MPEG::File& file) {
+#if (TAGLIB_HAS_TAG_CHECK)
+    return file.hasID3v2Tag();
+#else
+    return file.ID3v2Tag() != nullptr;
+#endif
+}
+
+bool hasAPETag(TagLib::MPEG::File& file) {
+#if (TAGLIB_HAS_TAG_CHECK)
+    return file.hasAPETag();
+#else
+    return file.APETag() != nullptr;
+#endif
+}
+
+bool hasID3v2Tag(TagLib::FLAC::File& file) {
+#if (TAGLIB_HAS_TAG_CHECK)
+    return file.hasID3v2Tag();
+#else
+    return file.ID3v2Tag() != nullptr;
+#endif
+}
+
+bool hasXiphComment(TagLib::FLAC::File& file) {
+#if (TAGLIB_HAS_TAG_CHECK)
+    return file.hasXiphComment();
+#else
+    return file.xiphComment() != nullptr;
+#endif
+}
+
+bool hasAPETag(TagLib::WavPack::File& file) {
+#if (TAGLIB_HAS_TAG_CHECK)
+    return file.hasAPETag();
+#else
+    return file.APETag() != nullptr;
+#endif
+}
+
+bool hasID3v2Tag(TagLib::RIFF::WAV::File& file) {
+#if (TAGLIB_HAS_WAV_ID3V2TAG)
+    return file.hasID3v2Tag();
+#else
+    return file.tag() != nullptr;
+#endif
+}
+
+bool hasID3v2Tag(TagLib::RIFF::AIFF::File& file) {
+#if (TAGLIB_HAS_AIFF_HAS_ID3V2TAG)
+    return file.hasID3v2Tag();
+#else
+    return file.tag() != nullptr;
+#endif
+}
+
+bool hasMP4Tag(TagLib::MP4::File& file) {
+    // Note (TagLib 1.11.1): For MP4 files without file tags
+    // TagLib still reports that the MP4 tag exists. Additionally
+    // we need to check that the tag itself is not empty.
+#if (TAGLIB_HAS_MP4TAG_CHECK_AND_IS_EMPTY)
+    return file.hasMP4Tag() && !file.tag()->isEmpty();
+#else
+    return file.tag() != nullptr;
+#endif
+}
+
+namespace {
```

Comment:
Personally, I think that anonymous namespace inside named namespaces are a bit confusing and prefer to put a single anonymous namespace at the top of the file. However, I guess it makes sense here if you don't want to clutter the code with namespace prefixes. But please move directly after the `namespace taglib {` line, not put in the middle of the code.